service/src/event/validator.rs

```diff
@@ -0,0 +1,269 @@
+
+use crate::Result;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum HistoryRequirement {
+    Complete,
+    InitOnly,
+    None,
+}
+
+impl From<DeliverableRequirement> for HistoryRequirement {
+    fn from(value: DeliverableRequirement) -> Self {
+        match value {
+            DeliverableRequirement::Immediate => HistoryRequirement::Complete,
+            DeliverableRequirement::Asap => HistoryRequirement::InitOnly,
+            DeliverableRequirement::Lazy => HistoryRequirement::None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EventValidator<'a> {
+    pool: &'a SqlitePool,
+    history_req: HistoryRequirement,
+}
+
+impl<'a> EventValidator<'a> {
+    pub fn new(pool: &'a SqlitePool, history_req: HistoryRequirement) -> Self {
+        Self { pool, history_req }
+    }
+
+    pub(crate) async fn validate_events(
+        &self,
+        parsed_events: Vec<EventInsertable>,
+        invalid_events: &mut Vec<InvalidItem>,
+    ) -> Result<(Vec<EventInsertable>, Vec<EventInsertable>)> {
+        // first collect all the init events so we can refer to them when grouping by type
+        // in order to avoid querying the database or rejecting the events incorrectly.
+        // we will then add any missing init events from the DB for the next phase.
+        // guess at a capacity and just assume half are init events so we only need to re-allocate once at most
+        // TODO: we should validate these init events before we trust them for others.. could
+        // somehow validate data as okay but init as invalid (???)
+        let mut required_init_events =
+            std::collections::HashMap::with_capacity(parsed_events.len() / 2);
+        parsed_events.iter().for_each(|event| {
+            if event.event().is_init() {
+                required_init_events.insert(*event.cid(), event.event().to_owned());
+            }
+        });
+        // Group events by their type (over allocate vectors for now)
+        let mut valid_events = Vec::with_capacity(parsed_events.len());
+        let mut signed_events = Vec::with_capacity(parsed_events.len());
+        let mut time_events = Vec::with_capacity(parsed_events.len());
+        let mut pending_events = match self.history_req {
+            HistoryRequirement::InitOnly => Vec::with_capacity(parsed_events.len() / 2),
+            _ => Vec::new(),
+        };
+        for event in parsed_events {
+            match event.event().as_ref() {
+                unvalidated::Event::Time(time) => {
+                    let init_known = if required_init_events.contains_key(time.id()) {
+                        true
+                    } else {
+                        let (exists, _) =
+                            CeramicOneEvent::deliverable_by_cid(self.pool, time.id()).await?;
+                        exists
+                    };
+                    if init_known || matches!(self.history_req, HistoryRequirement::None) {
```

Comment:
if there's no history requirement, then it seems like we can skip querying the database for the init event entirely yes?  Maybe we can restructure this with an early return to avoid doing unnecessary work?

Reply:
yeah, good call